kinesis/kinesis.go

```diff
@@ -29,6 +34,126 @@ const (
+// 2. If the buffer is not filled within 1 second, the data in the buffer will
+//    be flushed.
+// 4. If the calls to the Kinesis API become slow and fall behind, Writes will
+//    be dropped in 4mb chunks.
+func NewFastWriter(streamName, partitionKey string) *BufferedWriter {
+	return newFastWriter(NewDefaultWriter(streamName, partitionKey))
+}
+
+func newFastWriter(w io.Writer) *BufferedWriter {
+	// Allow up to 4 Writes to be queued, if the underlying io.Writer falls
+	// behind because of slow Writes, drop it.
+	dw := NewDropWriter(w, 4)
+
+	// Buffer 4mb in memory before flushing to Kinesis and periodically
+	// flush every second.
+	return NewBufferedWriter(dw, 4*1024*1024, time.Second)
+}
+
+type flushWriter interface {
+	io.Writer
+	Flush() error
+}
+
+// BufferedWriter is a wrapper around bufio.Writer that can periodically flush
+// the buffer instead of waiting for it to fill completely.
+type BufferedWriter struct {
+	w  flushWriter
+	mu sync.Mutex // Protects the underlying bufio.Writer, since we'll be flushing in a separate goroutine.
+	t  <-chan time.Time
+}
+
+// NewBufferedDefaultWriter returns a new io.Writer that buffers up to size
+// bytes in memory before flushing to the Kinesis. It will also periodically
+// flush the buffer if it is not full.
+func NewBufferedWriter(w io.Writer, size int, flush time.Duration) *BufferedWriter {
+	wr := &BufferedWriter{
+		w: bufio.NewWriterSize(w, size),
+		t: time.Tick(flush),
+	}
+	go wr.start()
+	return wr
+}
+
+// start receives on the ticker and flushes.
+func (w *BufferedWriter) start() {
+	for range w.t {
+		w.Flush()
+	}
+}
+
+// Write writes p to the underlying bufio.Writer.
+func (w *BufferedWriter) Write(p []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.w.Write(p)
+}
+
+// Flush flushe the underlying bufio.Writer.
+func (w *BufferedWriter) Flush() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.w.Flush()
+}
+
+// DropWriter is an io.Writer implementation that will drop writes when the
+// underlying io.Writer is too slow to process the write. It's generally
+// expected that you wrap this with a bufio.Writer.
+type DropWriter struct {
+	w       io.Writer
+	p       chan []byte
+	dropped func([]byte) // called when a Write is dropped.
+	err     error
+}
+
+// NewDropWriter returns a new DropWriter implementation that will drop writes.
+func NewDropWriter(w io.Writer, queue int) *DropWriter {
+	wr := &DropWriter{
+		w:       w,
+		dropped: Dropped,
+		p:       make(chan []byte, queue),
+	}
+	go wr.start()
+	return wr
+}
+
+// start starts writing the bytes in the queue.
+func (w *DropWriter) start() {
+	for p := range w.p {
+		if _, w.err = w.w.Write(p); w.err != nil {
+			return
+		}
+	}
+}
+
+// Write "writes" the bytes to the underlying channel. If the channel is full,
+// the packet is dropped.
+func (w *DropWriter) Write(p []byte) (int, error) {
+	if w.err != nil {
+		return 0, w.err
+	}
+
+	select {
+	case w.p <- p:
+	default:
+		w.dropped(p)
```

Comment:
Mike brought up a good point that it could be nice to write a message to the stream saying what was dropped.